### PR TITLE
[New Package] httpjson custom input

### PR DESCRIPTION
**Pull request by P1llus**
## What does this PR do?

This PR adds a package to utilize the httpjson input with custom integration, similar to winlog and log packages.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.13.0`).

### CODE REVIEW

**Comment by P1llus** [comment]
@ruflin @narph @andrewkroh 
Initial implementation of the httpjson input package, a few things I would love to get extra feedback on is:
1. Ensure the docs are written okay, and that the description of the manifest.yml items for the UI is good and understandable.
2. The title/description of the package itself.
3. I have added plenty of system tests, if you need any more just let me know.

This integration works very similarly to winlog and log packages, and is one of the many new input packages coming up. Started with this one as it is the most requested one at the moment.

File: packages/httpjson/_dev/build/docs/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Custom HTTPJSON input integration
+
+The custom HTTPJSON input integration is used to ingest data from custom RESTful API's that does
+not currently have an existing integration.
+
+The input itself supports sending both GET and POST requests, transform requests and responses during runtime, paginate and keep a running state on information from the last collected events.
+
+## Configuration
+
+The extensive documentation for the input is currently available [here](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html).
```
Comment (kaiyan-sheng) [inline packages/httpjson/_dev/build/docs/README.md:10]: Should we move all the documentations over into integrations so we don't have to link to Filebeat here?

Comment (P1llus) [reply packages/httpjson/_dev/build/docs/README.md:10]: Sure that's always a possibility as well, I was just unsure how large we would want the integration pages in fleet to be, as httpjson's documentation is large.

There is also lots of links to the httpjson documentation in the field descriptions of each option as well, but always up to discuss what is best :)

I personally think we should move it to live inside the integration as well.

**Comment by narph** [comment]
> @ruflin @narph @andrewkroh Initial implementation of the httpjson input package, a few things I would love to get extra feedback on is:
> 
> 1. Ensure the docs are written okay, and that the description of the manifest.yml items for the UI is good and understandable.
> 2. The title/description of the package itself.
> 3. I have added plenty of system tests, if you need any more just let me know.
> 4. There is currently no icon for this, the custom log input package uses a generic icon from our marketing or EUI styling, but there is nothing similar to match this usecase, any ideas?
> 

Code LGTM.

1. I would add the configuration options directly in the documentation, I guess if anything changes in Filebeat and is not available just yet in the integration might raise some confusion. 

4. Would the JSON icon do the trick here?

**Comment by P1llus** [comment]
@narph @kaiyan-sheng The place of the documentation is something we are still discussing internally, would you be able to see if there is anything else that should also change?
For example the wording of the UI field descriptions etc?
We might go ahead and merge this while discussing the decision on the documentation location, but that is still up to debate.

**Comment by P1llus** [comment]
For documentation it will most likely have to wait for: https://github.com/elastic/observability-docs/pull/424

File: packages/httpjson/data_stream/generic/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+          #  Param1:
+          #  - ValueA
+          #  - ValueB
+          #  Param2:
+          #    - Value
+      - name: response_decode_as
+        type: text
+        title: Response decode settings
+        description: |
+          ContentType used for decoding the response body. Supported values: application/json, application/x-ndjson. It is not set by default.
```
Comment (endorama) [inline packages/httpjson/data_stream/generic/manifest.yml:195]: I would specify how the response body is decoded in the default use case.

File: packages/httpjson/data_stream/generic/manifest.yml [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+      - name: oauth_token_url
+        type: text
+        title: Oauth2 Token URL
+        description: The URL endpoint that will be used to generate the tokens during the oauth2 flow. It is required if no provider is specified.
+        show_user: true
+        required: false
+      - name: request_body
+        type: yaml
+        title: Request Body
+        description: An optional HTTP body if the request method is POST
```
Comment (endorama) [inline packages/httpjson/data_stream/generic/manifest.yml:71]: Is there documentation to link here on how to build the request body?

File: packages/httpjson/data_stream/generic/manifest.yml [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+      - name: oauth_secret
+        type: password
+        title: Oauth2 Client Secret
+        description: Client secret used for Oauth2 authentication
+        show_user: true
+        required: false
+      - name: oauth_token_url
+        type: text
+        title: Oauth2 Token URL
+        description: The URL endpoint that will be used to generate the tokens during the oauth2 flow. It is required if no provider is specified.
```
Comment (endorama) [inline packages/httpjson/data_stream/generic/manifest.yml:65]: As there is no `provider` field, is unclear what this refers to. 

If it refers to `provider` property in `oauth_custom` I would phrase this differently, for example:
```suggestion
        description: The URL endpoint that will be used to generate the tokens during the oauth2 flow. It is required if no oauth_custom variable is set or provider is not specified in oauth_custom variable.
```

File: packages/httpjson/_dev/build/docs/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+The most commonly used configuration options is available on the main integration page, while more advanced and customizable options currently resides under the "advanced options" part of the integration settings page.
```
Comment (efd6) [inline packages/httpjson/_dev/build/docs/README.md:12]: s/is/are/
s/resides/reside/

File: packages/httpjson/_dev/build/docs/README.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+Configuration is split into 3 main categories, Request, Response and Cursor.
+
+The request part of the configuration handles points like which URL endpoint to communicate with, the request body, specific transformations that has to happen before a request is sent out and some custom options like request proxy, timeout and similar options.
```
Comment (efd6) [inline packages/httpjson/_dev/build/docs/README.md:16]: s/has/have/

File: packages/httpjson/_dev/build/docs/README.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# Custom HTTPJSON input integration
+
+The custom HTTPJSON input integration is used to ingest data from custom RESTful API's that does
```
Comment (efd6) [inline packages/httpjson/_dev/build/docs/README.md:3]: s/does/do/

File: packages/httpjson/data_stream/generic/agent/stream/httpjson.yml.hbs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+{{#if cursor}}
+cursor:
+  {{cursor}}
+{{/if}}
+
+{{#if processors}}
+processors:
+  {{processors}}
+{{/if}}
```
Comment (P1llus) [inline packages/httpjson/data_stream/generic/agent/stream/httpjson.yml.hbs:75]: Tags are missing

Comment (P1llus) [reply packages/httpjson/data_stream/generic/agent/stream/httpjson.yml.hbs:75]: There seems to already be a way to add ingest pipelines from all inputs as well, confirming this and will be adding that as well as an option

**Comment by P1llus** [comment]
Going ahead with the plans for these input packages, until the documentation has been moved to the new elastic-agent location, we will reference the filebeat documentation equivalent when needed.
Once documentation has been copied over to its new location, we will simply update the links to reference the new location.
Unless there is more reviews or questions around this, we will go ahead and merge this

**Comment by jamiehynds** [comment]
@andrewkroh could you please review this PR to ensure that we can publish the httpjson package (in line with the udp/tcp packages that shipped in 7.16). There some ongoing discussion on docs, but I don't think linking to Filebeat input docs should be a blocker. We can update the links once agent input docs are available (https://github.com/elastic/observability-docs/pull/424)

File: packages/httpjson/_dev/build/docs/README.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+The input itself supports sending both GET and POST requests, transform requests and responses during runtime, paginate and keep a running state on information from the last collected events.
+
+## Configuration
+
+The extensive documentation for the input are currently available [here](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html).
+
+The most commonly used configuration options are available on the main integration page, while more advanced and customizable options currently resides under the "advanced options" part of the integration settings page.
+
+Configuration is split into 3 main categories, Request, Response and Cursor.
```
Comment (andrewkroh) [inline packages/httpjson/_dev/build/docs/README.md:13]:
```suggestion
Configuration is split into three main categories, Request, Response and Cursor.
```

**Review by andrewkroh — APPROVED** [review]

**Pull request merged**